docs/using-modules.rst

```diff
@@ -0,0 +1,190 @@
+
+This is best illustrated with an example:
+
+.. code-block:: vyper
+
+    # ownable_2step.vy
+    import ownable
+
+    uses: ownable
+
+    # does not export ownable.transfer_ownership!
+
+    pending_owner: address  # the pending owner in the 2-step transfer process
+
+    @deploy
+    def __init__():
+        self.pending_owner = empty(address)
+
+    @external
+    def begin_transfer(new_owner: address):
+        ownable._check_owner()
+
+        self.pending_owner = new_owner
+
+    @external
+    def accept_transfer():
+        assert msg.sender == self.pending_owner
+
+        self.owner = new_owner
+        self.pending_owner = empty(address)
+
+Here, the ``ownable_2step`` module does not want to seal off access to calling the ``ownable`` module's ``__init__()`` function. So, it utilizes the ``uses: ownable`` statement to get access to the ``ownable`` module's state, without the requirement to initialize it. Note that this is a valid module, but it is not a valid contract (that is, it cannot produce bytecode) because it does not initialize the ``ownable`` module. To make a valid contract, the user of the ``ownable_2step`` module would be responsible for initializing the ``ownable`` module themselves (as in the next section: :ref:`initializing dependencies <init-dependencies>`).
+
+Whether to ``use`` or ``initialize`` a module is a choice which is left up to the library designer.
+
+Technical notes on the design
+-----------------------------
+
+This section contains some notes on the design from a language design perspective. It can be safely skipped if you are just interested in how to use modules, and not necessarily in programming language theory.
+
+The design of the module system takes inspiration from (but is not directly related to) the rust language's `borrow checker <https://doc.rust-lang.org/1.8.0/book/references-and-borrowing.html>`_. In the language of type systems, module initialization is modeled as an affine constraint which is promoted to a linear constraint if the module's state is touched in the compilation target. In practice, what this means is:
+
+* A module must be "used" or "initialized" before its state can be accessed in an import
+* A module may be "used" many times
+* A module which is "used" or its state touched must be "initialized" exactly once
+
+.. _init-dependencies:
+
+Initializing a module with dependencies
+=======================================
+
+Sometimes, you may encounter a module which itself ``uses`` other modules. Vyper's module system is designed to allow this, but it requires you make explicit the access to the imported module's state. The above ``ownable_2step.vy`` contract is an example of this. If you wanted to initialize the ``ownable_2step`` module, it would use the special ``:=`` syntax, and look something like this:
```

Comment:
```suggestion
Sometimes, you may encounter a module which itself ``uses`` other modules. Vyper's module system is designed to allow this, but it requires you make explicit the access to the imported module's state. The above ``ownable_2step.vy`` contract is an example of this. If you wanted to initialize the ``ownable_2step`` module, it would use the special ``:=`` (a.k.a. Walrus) syntax, and look something like this:
```